### PR TITLE
rusnapshot: add page

### PR DESCRIPTION
**Pull request by danyeet**
- [x] The page (if new), does not already exist in the repo.
- [x] The page is in the correct platform directory (`common/`, `linux/`, etc.)
- [x] The page has 8 or fewer examples.
- [x] The PR title conforms to the recommended [templates](/tldr-pages/tldr/blob/master/CONTRIBUTING.md#commit-message).
- [x] The page follows the [content guidelines](/tldr-pages/tldr/blob/master/CONTRIBUTING.md#guidelines).
- [x] The page description includes a link to documentation or a homepage (if applicable).

### CODE REVIEW

File: pages/linux/rusnapshot.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# rusnapshot
+
+> BTRFS snapshotting utility made in rust.
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:3]:
```suggestion
> BTRFS snapshotting utility made in Rust.
```

File: pages/linux/rusnapshot.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+> For more information: <https://github.com/Edu4rdSHL/rusnapshot>.
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:4]:
```suggestion
> More information: <https://github.com/Edu4rdSHL/rusnapshot>.
```

File: pages/linux/rusnapshot.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+- Create a snapshot using a config file:
+
+`sudo rusnapshot -c {{/path/to/config.toml}} --cr`
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:8]: Is there a long variant of `-c`, `-l`, or `-k`?

File: pages/linux/rusnapshot.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+- List created snapshots:
+
+`sudo rusnapshot -c {{/path/to/config.toml}} -l`
+
+- Delete snapshot by id or the name of the snapshot:
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:14]:
```suggestion
- Delete a snapshot by id or the name of the snapshot:
```

File: pages/linux/rusnapshot.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+`sudo rusnapshot -c {{/path/to/config.toml}} --del --id idprovidedbyrusnapshot`
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:16]:
```suggestion
`sudo rusnapshot -c {{/path/to/config.toml}} --del --id {{snapshot_id}}`
```

File: pages/linux/rusnapshot.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+- Delete all snapshots of [kind] three-hours:
```
Comment (marchersimon) [inline pages/linux/rusnapshot.md:18]: I don't really understand what you mean by `[kind]`. Is `three-hours` a special kind of snapshot?

**Review by marchersimon — CHANGES_REQUESTED** [review]
Thanks @danyeet. I left a few suggestions below. Also, if `rusnapshot` also works with relative paths, please delete all leading slashes.

File: pages/linux/rusnapshot.md (hunk 3 above)
Comment (danyeet) [reply pages/linux/rusnapshot.md:8]: Yes, for -l is --list, for -k is --keep and for -c is --config.

File: pages/linux/rusnapshot.md (hunk 6 above)
Comment (Edu4rdSHL) [reply pages/linux/rusnapshot.md:18]: Hi, rusnapshot dev here. `kind` is just an identifier for the snapshot, you can put any string that you want. Why it exists? Well, the tool is able to perform automatic cleanups of snapshots, you can use the `--keep` option to keep only the last X snapshots, but see the following scenario: I took hourly, weekly and monthly snapshots of the same subvolumes, without an identifier like `--kind` it's almost impossible to difference between them and the tool will treat all the snapshots equally. To solve this problem you can use  `--kind hourly` for the hourly snapshots, `--kind weekly` for the weekly snapshots and `--kind montly` for the monthly snapshots, that way when the tool performs the cleanups, it's able to "see" that there are three different "kind" of snapshots and that only need to clean the ones that match the current "kind", leaving the others intact.

I hope that I have been clear, let me know if you have any additional question.

Regards,
Ed

**Comment by Edu4rdSHL** [comment]
Regarding to slashes, we are talking about btrfs subvolumes. The tool can actually works with relative paths but it's **highly** recommended to use full qualified paths, the reason is that the tool is also able to restore snapshots, so if you use relative paths and then try restoring the snapshot it will ends up creating the subvolume in the current directory what may or may not be correct. If the user wants to restore to an specific directory he can use `--from {{desired-directory-path}}` while using `--restore`.

So, I prefer to keep the full qualified paths.

Regards,
Ed

File: pages/linux/rusnapshot.md (hunk 3 above)
Comment (bl-ue) [reply pages/linux/rusnapshot.md:8]: Nice! then @danyeet, let's replace the short options with the long options.  In case you don't know, we prefer the long options to the short ones because since they're normally English words, they're essentially self-explanatory, thus providing documentation in addition to that in the example description.  Makes sense?
 ```suggestion
`sudo rusnapshot --config {{/path/to/config.toml}} --cr`
```

**Comment by bl-ue** [comment]
Hi @Edu4rdSHL, thank you very much for your input!  I'm perfectly fine with that—if the most common usages of a command are with absolute paths, then by all means that's what we'll do.

What about the path to the config file, though? In this page, it's `/path/to/config.toml`, which is absolute—but normally that type of thing works fine with being relative.  Is that the case here, or is it advisable to use an absolute path even for the config file?

File: pages/linux/rusnapshot.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# rusnapshot
+
+> BTRFS snapshotting utility made in Rust.
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:3]: I like prefer the verb `written` to `made` when discussing the software stack.  Cool that it's in Rust! 🚀
```suggestion
> BTRFS snapshotting utility written in Rust.
```

File: pages/linux/rusnapshot.md [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+> More information: <https://github.com/Edu4rdSHL/rusnapshot>.
+
+- Create a snapshot using a config file:
+
+`sudo rusnapshot -c {{/path/to/config.toml}} --cr`
+
+- List created snapshots:
+
+`sudo rusnapshot -c {{/path/to/config.toml}} -l`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:12]:
```suggestion
`sudo rusnapshot --config {{/path/to/config.toml}} --list`
```

File: pages/linux/rusnapshot.md (hunk 3 above)
Comment (danyeet) [reply pages/linux/rusnapshot.md:8]: Yeah, fully agree, just to be honest I had to double check if there were long options in the first place :D.

File: pages/linux/rusnapshot.md [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+- Delete a snapshot by ID or the name of the snapshot:
+
+`sudo rusnapshot -c {{/path/to/config.toml}} --del --id {{snapshot_id}}`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:16]:
```suggestion
`sudo rusnapshot --config {{/path/to/config.toml}} --del --id {{snapshot_id}}`
```

File: pages/linux/rusnapshot.md [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+- Delete all snapshots of [kind] three-hours:
+
+`sudo rusnapshot -c {{/path/to/config.toml}}  -l -k 0 --clean --kind three-hours`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:20]:
```suggestion
`sudo rusnapshot --config {{/path/to/config.toml}}  --list --keep 0 --clean --kind {{three-hours}}`
```

File: pages/linux/rusnapshot.md (hunk 7 above)
Comment (danyeet) [reply pages/linux/rusnapshot.md:3]: I'll fix it right at it.
Also yeah rust is very cool.

File: pages/linux/rusnapshot.md (hunk 10 above)
Comment (danyeet) [reply pages/linux/rusnapshot.md]: I think I made most of the changes 
I'll add the brackets for the variable

Comment (danyeet) [reply pages/linux/rusnapshot.md]: Please take a look again if you find something I'll edit it right at it but I think for the most part is fine.

File: pages/linux/rusnapshot.md [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --list`
+
+- Delete a snapshot by ID or the name of the snapshot:
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --del --id {{snapshot_id}}`
+
+- Delete all snapshots of [kind] three-hours:
+
+`sudo rusnapshot --config {{/path/to/config.toml}}  --list --keep 0 --clean --kind {{three-hours}}`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:20]:
```suggestion
`sudo rusnapshot --config {{/path/to/config.toml}}  --list --keep {{0}} --clean --kind {{three-hours}}`
```

File: pages/linux/rusnapshot.md (hunk 6 above)
Comment (bl-ue) [reply pages/linux/rusnapshot.md:18]: Maybe we should call it something like `hourly` then.  @danyeet what do you think?

Comment (danyeet) [reply pages/linux/rusnapshot.md:18]: No problem with that - the idea is to show that it's a time interval so I'll change it no problem.

Comment (marchersimon) [reply pages/linux/rusnapshot.md]: Also I'd rephrase the description a little. Maybe `- Delete all hourly snapshots:`.

Comment (danyeet) [reply pages/linux/rusnapshot.md]: Yea that would be better. Just to clarify something - the time interval convention is just an example. You could maybe make something else, like backup before update, backup before restart and etc. We just thought it would be the easiest one to explain.

Comment (bl-ue) [reply pages/linux/rusnapshot.md]: I'm inclined to prefer the first (`Delete all snapshots of kind hourly`) because `hourly` is just a name, with no meaning otherwise.  It could easily be `abc`, whereupon your suggestion @marchersimon would not make as much sense (`Delete all abc snapshots:`)

Comment (danyeet) [reply pages/linux/rusnapshot.md]: Maybe just reformat hourly to [hourly] or {{hourly}} ? 
I mean I can see both sides, I am fine with both but abstracting it a bit might not be a bad idea :).

Comment (marchersimon) [reply pages/linux/rusnapshot.md]: > the time interval convention is just an example

Ah, I tought it was all about time intervals. In that case `Delete all snapshots of kind hourly` is more universal. However in descriptions we use `` `hourly` ``, instead of `{{hourly}}` for variables.

Comment (danyeet) [reply pages/linux/rusnapshot.md]: Gotcha.

**Comment by Edu4rdSHL** [comment]
Hey @bl-ue, sorry for the delay. Yes, the path to the configuration file can be relative without any problem and doesn't really matter if it's relative/full.

File: pages/linux/rusnapshot.md (hunk 6 above)
Comment (Edu4rdSHL) [reply pages/linux/rusnapshot.md]: I have to clarify that the `--kind` option is not just for deleting, the deletion only happens if you use the `--clean` and `--keep` but the option can be used while creating new snapshots without the need of deleting previous ones. As I said initially, it's more of an identifier than anything else.

File: pages/linux/rusnapshot.md [hunk 12]
```diff
@@ -0,0 +1,20 @@
+# rusnapshot
+
+> BTRFS snapshotting utility written in Rust.
+> More information: <https://github.com/Edu4rdSHL/rusnapshot>.
+
+- Create a snapshot using a config file:
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --cr`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:8]:
```suggestion
`sudo rusnapshot --config {{path/to/config.toml}} --cr`
```

File: pages/linux/rusnapshot.md [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+
+- List created snapshots:
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --list`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:12]:
```suggestion
`sudo rusnapshot --config {{path/to/config.toml}} --list`
```

File: pages/linux/rusnapshot.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+- Delete a snapshot by ID or the name of the snapshot:
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --del --id {{snapshot_id}}`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:16]:
```suggestion
`sudo rusnapshot --config {{path/to/config.toml}} --del --id {{snapshot_id}}`
```

File: pages/linux/rusnapshot.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+
+- Delete all `hourly` snapshots:
+
+`sudo rusnapshot --config {{/path/to/config.toml}} --list --keep {{0}} --clean --kind {{hourly}}`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:20]:
```suggestion
`sudo rusnapshot --config {{path/to/config.toml}} --list --keep {{0}} --clean --kind {{hourly}}`
```

**Comment by danyeet** [comment]
I'll add two more examples as @Edu4rdSHL and I think are crucial.

**Review by bl-ue — DISMISSED** [review]
Thank you @danyeet (and @Edu4rdSHL)! 🚀

**Comment by danyeet** [comment]
After messing up three times in a row, I have added two more examples. So the first one is creating a read-write snapshot as the utility creates a read-only by default. The second one is restoring the created snapshot to a subvolume. `--from` specifies where.

**Review by marchersimon — APPROVED** [review]

**Review by bl-ue — APPROVED** [review]

**Comment by danyeet** [comment]
Thank you for working it out with me and @Edu4rdSHL - it's great we can popularize this rust piece of art already haha.

**Comment by Edu4rdSHL** [comment]
We should clarify that `--from` is only required if you want to restore the snaphsot to an specific place and not the one from where the snapshot was originally taken. If only `--restore --id {{ID}}` is provided, it will restore the snapshot to the original directory.

**Comment by danyeet** [comment]
Since in the last example `--from` is not a strict requirement, should we omit it, or it's fine as is ?

**Comment by Edu4rdSHL** [comment]
My vote is that we must remove `--from` from the restoration example as it's only required under very specific scenarios.

**Review by Edu4rdSHL — APPROVED** [review]

**Review by marchersimon — APPROVED** [review]

File: pages/linux/rusnapshot.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+`sudo rusnapshot --config {{path/to/config.toml}} --list --keep {{0}} --clean --kind {{hourly}}`
+
+- Create a read-write snapshot:
+
+`sudo rusnapshot --config {{path/to/config.toml}} --cr --rw`
+
+- Restore a snapshot:
+
+`sudo rusnapshot --config {{path/to/config.toml}} --id {{snapshot_id}} --restore`
```
Comment (bl-ue) [inline pages/linux/rusnapshot.md:28]: Since the `--config` is always present, how about we clear it a little and shorten all but the first occurence?

```suggestion
`sudo rusnapshot -c {{path/to/config.toml}} --list`

- Delete a snapshot by ID or the name of the snapshot:

`sudo rusnapshot -c {{path/to/config.toml}} --del --id {{snapshot_id}}`

- Delete all `hourly` snapshots:

`sudo rusnapshot -c {{path/to/config.toml}} --list --keep {{0}} --clean --kind {{hourly}}`

- Create a read-write snapshot:

`sudo rusnapshot -c {{path/to/config.toml}} --cr --rw`

- Restore a snapshot:

`sudo rusnapshot -c {{path/to/config.toml}} --id {{snapshot_id}} --restore`
```

**Comment by danyeet** [comment]
Well definitely it looks way nicer. Testing it locally with tealdeer - looks very nice

**Review by bl-ue — APPROVED** [review]

**Review by Edu4rdSHL — APPROVED** [review]
Thank you @danyeet for creating the PR and thanks to @bl-ue & @marchersimon for the reviews, LGTM.

**Comment by danyeet** [comment]
Thank you very much. Look forward to more contributions from me in the future. :rocket:
